From aabcd93a2dd7d272f9aeb4e7b65b7eec57f5eea2 Mon Sep 17 00:00:00 2001
From: Ricardo Pardini <ricardo@pardini.net>
Date: Thu, 26 Dec 2024 19:00:28 +0100
Subject: [PATCH] [userpatches/rockchip64-6.18/dt] cm3588-nas: extra-DT for
 (Noctua...) 4-pin pwm 5v fan on 40-pin pwm5 (drive) & pwm9 (tacho)

- Add a kernel cooling map tripping at 60C - needs more
- I _dare you_ (yes, *you!*) make an overlay out of this ;-)
 - cm3588-nas-40pin-pwm-fan: more trips from 50 to 70C

40pin-pwm-fan MCE 6.18
---
 ...-friendlyelec-cm3588-nas-40pin-pwm-fan.dts | 144 ++++++++++++++++++
 1 file changed, 144 insertions(+)
 create mode 100644 userpatches/kernel/archive/rockchip64-6.18/dt/rk3588-friendlyelec-cm3588-nas-40pin-pwm-fan.dts

diff --git a/arch/arm64/boot/dts/rockchip/rk3588-friendlyelec-cm3588-nas-40pin-pwm-fan.dts b/arch/arm64/boot/dts/rockchip/rk3588-friendlyelec-cm3588-nas-40pin-pwm-fan.dts
new file mode 100644
index 0000000000..b714b61659
--- /dev/null
+++ b/arch/arm64/boot/dts/rockchip/rk3588-friendlyelec-cm3588-nas-40pin-pwm-fan.dts
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+
+/dts-v1/;
+
+#include "rk3588-friendlyelec-cm3588-nas.dts"
+
+/ {
+	model = "FriendlyElec CM3588 NAS with 4-pin 5v PWM fan (drive pin 32/tacho pin 33) plus MCE IR receiver";
+
+	fan: pwm-fan {
+		compatible = "pwm-fan";
+		#cooling-cells = <2>;
+		cooling-levels = <0 26 51 77 102 128 153 179 204 230 255>;
+		pwms = <&pwm5 0 50000 0>; /* GPIO 40-pin Connector pin 32 - and NOT the original pwm1 which MOSFET-drives the built-in small-connector "5v fan" */
+		fan-supply = <&vcc_5v0_sys>;
+		pulses-per-revolution = <2>;
+		interrupt-parent = <&gpio3>;  /* GPIO GPIO3_B0/PWM9_M0 + ... */
+		interrupts = <RK_PB0 IRQ_TYPE_EDGE_FALLING>; /* ... + GPIO 40-pin Connector pin 33 */
+		status = "okay";
+	};
+
+	// rpardini abusing this for unrelated goodness
+	ir-receiver {
+		linux,rc-map-name = "rc-rc6-mce";
+		wakeup-source;
+	};
+
+};
+
+/* Temperature sensor near the center of the SoC */
+&package_thermal {
+	polling-delay = <1000>;
+
+	trips {
+		trip0: trip0 {
+			hysteresis = <2000>;
+			temperature = <50000>; /* 50 celsius */
+			type = "active";
+		};
+		trip1: trip1 {
+			hysteresis = <2000>;
+			temperature = <52000>; /* 52 celsius */
+			type = "active";
+		};
+		trip2: trip2 {
+			hysteresis = <2000>;
+			temperature = <54000>; /* 54 celsius */
+			type = "active";
+		};
+		trip3: trip3 {
+			hysteresis = <2000>;
+			temperature = <56000>; /* 56 celsius */
+			type = "active";
+		};
+		trip4: trip4 {
+			hysteresis = <2000>;
+			temperature = <58000>; /* 58 celsius */
+			type = "active";
+		};
+		trip5: trip5 {
+			hysteresis = <2000>;
+			temperature = <60000>; /* 60 celsius */
+			type = "active";
+		};
+		trip6: trip6 {
+			hysteresis = <2000>;
+			temperature = <62000>; /* 62 celsius */
+			type = "active";
+		};
+		trip7: trip7 {
+			hysteresis = <2000>;
+			temperature = <64000>; /* 64 celsius */
+			type = "active";
+		};
+		trip8: trip8 {
+			hysteresis = <2000>;
+			temperature = <66000>; /* 66 celsius */
+			type = "active";
+		};
+		trip9: trip9 {
+			hysteresis = <2000>;
+			temperature = <70000>; /* 70 celsius */
+			type = "active";
+		};
+	};
+
+	cooling-maps {
+		map0 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip0>;
+		};
+		map1 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip1>;
+		};
+		map2 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip2>;
+		};
+		map3 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip3>;
+		};
+		map4 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip4>;
+		};
+		map5 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip5>;
+		};
+		map6 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip6>;
+		};
+		map7 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip7>;
+		};
+		map8 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip8>;
+		};
+		map9 {
+			cooling-device = <&fan THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+			trip = <&trip9>;
+		};
+	};
+};
+
+/* GPIO 40-pin Connector pin 33 - EXTFAN TACHO - used as interrupt source in pwm-fan - gpio GPIO3_B0/PWM9_M0 */
+&pwm9 {
+	pinctrl-0 = <&pwm9m0_pins>;
+	status = "disabled"; /* unsure if this is enabled or disabled, as it is used as an interrupt source and not pwm output */
+};
+
+/* Just for reference, as it is already okay'ed on upstream DT (wasn't on Nargas) */
+/* GPIO 40-pin Connector pin 32 - EXTFAN PWM Drive */
+/* &pwm5 {
+	pinctrl-0 = <&pwm5m1_pins>;
+	status = "okay";
+};
+*/
+
--
2.52.0

